File: runtime/src/test/py/annotation_page.py
#
# microtome

from microtome.page import Page
from microtome.core.prop import Prop, PropSpec
from primitive_page import PrimitivePage
import types

class AnnotationPage(Page):
    def __init__(self, name):
        super(AnnotationPage, self).__init__(name)
        self._foo = Prop(self, AnnotationPage._s_fooSpec)
        self._bar = Prop(self, AnnotationPage._s_barSpec)
        self._primitives = Prop(self, AnnotationPage._s_primitivesSpec)

    @property
    def foo(self):
        return self._foo.value

    @property
    def bar(self):
        return self._bar.value

    @property
    def primitives(self):
        return self._primitives.value

    @property
    def props(self):
        return super(AnnotationPage, self).props + [self._foo, self._bar, self._primitives]

    _s_fooSpec = PropSpec("foo", {"min": 3.0, "max": 5.0}, [types.IntType, ])
    _s_barSpec = PropSpec("bar", {"default": 3.0}, [types.IntType, ])
    _s_primitivesSpec = PropSpec("primitives", {"nullable": True}, [PrimitivePage, ])

if __name__ == "__main__":
    print AnnotationPage._s_barSpec._annotations

File: runtimes/objc/Tests/AnnotationPage.m

#import "AnnotationPage.h"
#import "PrimitivePage.h"
#import "microtome.h"

static MTPropSpec* s_fooSpec = nil;
static MTPropSpec* s_barSpec = nil;
static MTPropSpec* s_primitivesSpec = nil;

@implementation AnnotationPage {
@protected
    MTIntProp* _foo;
    MTIntProp* _bar;
    MTObjectProp* _primitives;
}

- (int)foo { return _foo.value; }
- (int)bar { return _bar.value; }
- (PrimitivePage*)primitives { return _primitives.value; }

- (NSArray*)props { return MT_PROPS(_foo, _bar, _primitives, ); }

- (id)init {
    if ((self = [super init])) {
        _foo = [[MTIntProp alloc] initWithPropSpec:s_fooSpec];
        _bar = [[MTIntProp alloc] initWithPropSpec:s_barSpec];
        _primitives = [[MTObjectProp alloc] initWithPropSpec:s_primitivesSpec];
    }
    return self;
}

+ (void)initialize {
    if (self == [AnnotationPage class]) {
        s_fooSpec = [[MTPropSpec alloc] initWithName:@"foo"
            annotations:@{ @"min":@3.0,@"max":@5.0, }
            valueClasses:nil ];
        s_barSpec = [[MTPropSpec alloc] initWithName:@"bar"
            annotations:@{ @"default":@3.0, }
            valueClasses:nil ];
        s_primitivesSpec = [[MTPropSpec alloc] initWithName:@"primitives"
            annotations:@{ @"nullable":@YES, }
            valueClasses:@[ [PrimitivePage class], ] ];
    }
}

@end
